packages/simulation-new/src/actors/Plant.ts

```diff
@@ -102,84 +88,32 @@ export class Plant {
-    this.actyx.subscribe(
-      {
-        query: PlantHappenings.TagPlantWatered,
-      },
-      (e) => {
-        if (e.meta.lamport < this.initializationLamportTime) {
-          return;
-        }
-
-        this.water += 30;
+    PlantHappenings.subscribeWaterEventById(
+      this.actyx,
+      this.data.id,
+      (meta) => {
+        console.log("water");
+        if (meta.lamport < this.initializationLamportTime) return;
+        PlantData.WaterLevel.applyWater(this.data);
       }
     );
   }
 
-  private static initId(): string {
-    let plantId = localStorage.getItem("plantId");
-    if (!plantId) {
-      plantId = v4();
-      localStorage.setItem("plantId", plantId);
-    }
-    return plantId;
-  }
-
-  private static async retrieveIdFromActyx(
-    id: string,
-    actyx: Actyx
-  ): Promise<{ data: Sensor.Type; lamport: number } | undefined> {
-    const actyxEvents = await actyx.queryAql({
-      query: `FROM ${WorldCreateWithId(id)} `,
-    });
-    const event = actyxEvents
-      .filter((e): e is AqlEventMessage => e.type === "event")
-      .at(0);
-    const parsed = Sensor.Type.safeParse(event?.payload);
-    if (parsed.success) {
-      return { data: parsed.data, lamport: event?.meta.lamport || 0 };
-    }
-    return undefined;
-  }
-
-  private static async retrieveLatestStateFromActyx(
-    id: string,
-    actyx: Actyx
-  ): Promise<{ data: number; lamport: number } | undefined> {
-    const actyxEvents = await actyx.queryAql({
-      query: `
-                PRAGMA features := aggregate
-                FROM ${PlantHappenings.TagPlantWithId(id)} & ${WorldUpdate}
-                AGGREGATE LAST(_.water)
-            `,
-    });
-    const event = actyxEvents
-      .filter((e): e is AqlEventMessage => e.type === "event")
-      .at(0);
-    const latestWaterValue = event?.payload as number; // should be safe due to the kind of query
-    return { data: latestWaterValue, lamport: event?.meta.lamport || 0 };
-  }
-
   private measureWater(): number {
     const currentMeasurement = Date.now();
     const elapsed = currentMeasurement - this.lastMeasurement;
-    const drainedWaterAmount = (elapsed / 1000) * WATER_DRAIN
-    const newWaterAmount = this.water - drainedWaterAmount
+    const drainedWaterAmount = (elapsed / 1000) * WATER_DRAIN;
+    const newWaterAmount = this.data.water - drainedWaterAmount;
     // rounds to 2 decimal places
-    const roundedWaterLevel = Math.round(newWaterAmount * 100) / 100
+    const roundedWaterLevel = Math.round(newWaterAmount * 100) / 100;
 
-    this.water = Math.max(roundedWaterLevel, 0);
+    this.data.water = Math.max(roundedWaterLevel, 0);
     this.lastMeasurement = currentMeasurement;
 
-    return this.water;
+    return this.data.water;
   }
 
-  private toPayload(): Sensor.Type {
-    return {
-      id: this.id,
-      water: this.water,
-      pos: this.position,
-      t: "Sensor",
-    };
+  private getData(): PlantData.Type {
+    return { ...this.data };
```

Comment:
Doesn't TS have the "Kotlin/Swift style getters"? 

Like `get thisSuperCoolVar: number` or something?

Reply:
returning `readonly` is a thing, a possible alternative for this, but then the `readonly` marked object can't be passed around to functions that requires `non-readonly`